amr-wind/wind_energy/ABLCodedInlet.cpp

```diff
@@ -0,0 +1,200 @@
+
+        for (amrex::MFIter mfi(mfab); mfi.isValid(); ++mfi) {
+            auto gbx = amrex::grow(mfi.validbox(), nghost);
+            if (!gbx.cellCentered()) {
+                gbx.enclosedCells();
+            }
+            const auto& bx = gbx & dbx;
+            if (!bx.ok()) {
+                continue;
+            }
+
+            const auto& arr = mfab[mfi].array();
+            const int numcomp = mfab.nComp();
+
+            amrex::ParallelFor(
+                bx, [=, coded = m_user_coded] AMREX_GPU_DEVICE(
+                        int i, int j, int k) noexcept {
+                    const amrex::Real x = problo[0] + (i + 0.5) * dx[0];
+                    const amrex::Real y = problo[1] + (j + 0.5) * dx[1];
+                    const amrex::Real z = problo[2] + (k + 0.5) * dx[2];
+
+                    double Vtmp[3];
+                    coded.get_vel(time, x, y, z, Vtmp);
+
+                    amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> vels = {
+                        AMREX_D_DECL(Vtmp[0], Vtmp[1], Vtmp[2])};
+                    for (int n = 0; n < numcomp; n++) {
+                        arr(i, j, k, dcomp + n) = vels[orig_comp + n];
+                    }
+                });
+        }
+    }
+}
+
+void ABLCodedInlet::set_temperature(
+    const int lev,
+    const amrex::Real time,
+    const Field& fld,
+    amrex::MultiFab& mfab) const
+{
+    if (!m_active) {
+        return;
+    }
+
+    BL_PROFILE("amr-wind::ABLCodedInlet::set_temperature");
+
+    const auto& geom = m_mesh.Geom(lev);
+    const auto& problo = geom.ProbLoArray();
+    const auto& dx = geom.CellSizeArray();
+
+    const auto& bctype = fld.bc_type();
+    const int nghost = 1;
+    const auto& domain = geom.growPeriodicDomain(nghost);
+
+    for (amrex::OrientationIter oit; oit != nullptr; ++oit) {
+        auto ori = oit();
+        if (bctype[ori] != BC::mass_inflow) {
+            continue;
+        }
+
+        const int idir = ori.coordDir();
+        const auto& dbx = ori.isLow() ? amrex::adjCellLo(domain, idir, nghost)
+                                      : amrex::adjCellHi(domain, idir, nghost);
+
+        for (amrex::MFIter mfi(mfab); mfi.isValid(); ++mfi) {
+            auto gbx = amrex::grow(mfi.validbox(), nghost);
+            if (!gbx.cellCentered()) {
+                gbx.enclosedCells();
+            }
+            const auto& bx = gbx & dbx;
+            if (!bx.ok()) {
+                continue;
+            }
+
+            const auto& arr = mfab[mfi].array();
+
+            amrex::ParallelFor(
+                bx, [=, coded = m_user_coded] AMREX_GPU_DEVICE(
+                        int i, int j, int k) noexcept {
+                    const amrex::Real x = problo[0] + (i + 0.5) * dx[0];
+                    const amrex::Real y = problo[1] + (j + 0.5) * dx[1];
+                    const amrex::Real z = problo[2] + (k + 0.5) * dx[2];
+
+                    coded.get_temp(time, x, y, z, arr(i, j, k));
```

Comment:
So apparently SYCL kernel cannot call through a function pointer. @marchdf Do you know of workarounds here?

Reply:
I will add though, I don't understand the design here. Using a function pointer makes little sense to me, considering that it cannot represent both the host and device version.

Reply:
The function pointers were new for me. But it was based on examples I found for implementing runtime library linking. Perhaps there's another way?

Reply:
If there isn't a cross-platform solution, I could just create a cmake flag for this that's disabled by default...

Reply:
hmmm cross-platform is going to be a problem I think... This is a hard thing to accomplish. In general, if you are going to compile a file to be included as a dl, why not just edit the code? We could make the use/definition of that kind of function easier for the user?

Reply:
@marchdf I see your point and that was going to be my original approach. But I was curious to see if this approach would work because it (arguably) has more usability and flexibility:
* The user can start with a pre-compiled amr-wind executable (that is, e.g., set up to be performant for the target system) and develop custom field functions w/o having to compile amr-wind on their own. (Side note: From tiny profiler, the `set_velocity` calls from this solution path amounts to 0.03%)
* The user will not have to understand the nuances of AMReX or GPU programming.
* This allows general inflows to be specified and is comparable to the OpenFOAM's codedFixedValue BC.

Reply:
Yeah I see the utility of this. Esp since OpenFoam does something like it. But (if I recall?) I think you supply text to openfoam and it converts it to code? There's also the option of driving this through py-amrex and then we can provide an easy python UDF...

Reply:
OpenFOAM has a JIT compiler setup, so this solution comes up a little bit short in comparison. I'm not sure how to implement py-amrex -- do you?

Reply:
Not off the top of my head. But you would need a py-amrex to drive the sim and then hooks in amr-wind to have the right callbacks. It would be work...